rust/agama-dbus-server/src/main.rs

```diff
@@ -0,0 +1,109 @@
+pub mod error;
+use crate::error::Error;
+
+use std::{future::pending, process::Command};
+use agama_locale_data::{parse_locale};
+use anyhow::Context;
+use zbus::{ConnectionBuilder, dbus_interface};
+
+struct Locale {
+    locale: String,
+    keymap: String,
+    timezone_id: String
+}
+
+#[dbus_interface(name = "org.opensuse.Agama.Locale1")]
+impl Locale {
+    // Can be `async` as well.
+    /// get labels for given locale. The first pair is english language and territory
+    /// and second one is localized one to target language from locale.
+    /// 
+    fn labels_for_locale(&self, locale: &str) -> Result<((String, String), (String, String)), Error> {
+        const DEFAULT_LANG : &str = "en";
+        let (loc_language, loc_territory) = parse_locale(locale)?;
+        let languages = agama_locale_data::get_languages()?;
+        let territories = agama_locale_data::get_territories()?;
+        let language = languages.language.iter()
+            .find(|l| l.id == loc_language).context("language for passed locale not found")?;
+        let territory = territories.territory.iter()
+            .find(|t| t.id == loc_territory).context("territory for passed locale not found")?;
+
+        let default_ret = (language.names.name_for(DEFAULT_LANG).context("missing default translation for language")?,
+            territory.names.name_for(DEFAULT_LANG).context("missing default translation for territory")?);
+        let localized_ret = (language.names.name_for(language.id.as_str()).context("missing default translation for language")?,
+            territory.names.name_for(language.id.as_str()).context("missing default translation for territory")?);
+        Ok((default_ret, localized_ret))        
+    }
+
+    fn set_locale(&mut self, locale: &str) {
```

Comment:
question is where to check for invalid values. Here or during commit? Also I am not sure if we even can check properly all locales, but we can somehow construct list of it from langtable I expect. or from `localectl list-locales` but it can be limited....or third option which is probably best is to have setter for supported locales like `set_locales` that software uses and then we get return labels without input parameter. What do you think?

Reply:
As a D-Bus user, I would like to check the value when it is assigned (the sooner the better). Otherwise, the service might contain some invalid state.